ophyd/signal.py

```diff
@@ -1182,58 +1209,65 @@ def _pv_access_callback(self, read_access, write_access, pv):
             md_update['write_access'] = write_access
 
         if md_update:
             self._metadata.update(**md_update)
 
         if self.connected:
-            self._run_subs(sub_type=self.SUB_META, **self._metadata)
+            self._run_metadata_callbacks()
 
         super()._pv_access_callback(read_access, write_access, pv)
         self._set_event_if_ready()
 
-    def _metadata_changed(self, pvname, cl_metadata, *, require_timestamp=False,
-                          update=True):
+    def _metadata_changed(self, pvname, cl_metadata, *, from_monitor,
+                          require_timestamp=False, update=True):
         'Metadata for one PV has changed'
         metadata = super()._metadata_changed(
-            pvname, cl_metadata, update=update,
+            pvname, cl_metadata, from_monitor=from_monitor, update=update,
             require_timestamp=require_timestamp)
 
-        if (self.setpoint_pvname != self._read_pvname and
-                self.setpoint_pvname == pvname):
-            self._run_subs(sub_type=self.SUB_SETPOINT_META,
-                           timestamp=self._metadata['setpoint_timestamp'],
-                           status=self._metadata['setpoint_status'],
-                           severity=self._metadata['setpoint_severity'],
-                           precision=self._metadata['setpoint_precision'],
-                           lower_ctrl_limit=self._metadata['lower_ctrl_limit'],
-                           upper_ctrl_limit=self._metadata['upper_ctrl_limit'],
-                           units=self._metadata['units'],
-                           )
+        if all((from_monitor,
+                self.setpoint_pvname != self.pvname,
+                self.setpoint_pvname == pvname)):
+            # Setpoint has its own metadata
+            self._metadata_thread_ctx.run(
+                self._run_subs,
+                sub_type=self.SUB_SETPOINT_META,
+                timestamp=self._metadata['setpoint_timestamp'],
+                status=self._metadata['setpoint_status'],
+                severity=self._metadata['setpoint_severity'],
+                precision=self._metadata['setpoint_precision'],
+                lower_ctrl_limit=self._metadata['lower_ctrl_limit'],
+                upper_ctrl_limit=self._metadata['upper_ctrl_limit'],
+                units=self._metadata['units'],
+            )
         return metadata
 
     def _write_changed(self, value=None, timestamp=None, **kwargs):
-        '''A callback indicating that the write value has changed'''
+        'CA monitor: callback indicating the setpoint PV value has changed'
         if timestamp is None:
             timestamp = time.time()
 
+        metadata = self._metadata_changed(self.setpoint_pvname, kwargs,
+                                          require_timestamp=True,
+                                          from_monitor=True,
+                                          update=False)
```

Comment:
@klauer left a comment suggesting changing to `update=True` was a good idea and resolving the conversation, but this is still `update=False`.

Reply:
Sloppy, @klauer, sloppy! Addressed in a follow-up commit.